### PR TITLE
Cp Target Normalization: Re-invariant pressure coefficient targets

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Currently the model predicts raw pressure p (after asinh transformation). But the physically invariant quantity in external aerodynamics is the **pressure coefficient**: `Cp = (p - p_inf) / (0.5 * rho * Umag^2)`. Cp is dimensionless and approximately **Re-independent** for attached flow. By predicting Cp instead of raw p, the model learns a Re-invariant target, which should improve **p_re** (OOD Reynolds) generalization.

**Why this should work:** The p_re regression (6.300 → 6.364 from PR #2213 to #2251) indicates the model struggles with Reynolds-number-dependent pressure scaling. Cp normalization directly removes the Re² scaling from pressure targets. This is the theoretically correct normalization for Re-invariant pressure prediction.

**Key difference from prior Cp work (PR #392, Phase 1):** That early experiment tested Cp normalization before residual prediction, SRF heads, PCGrad, or asinh transform existed. The question is whether Cp normalization on top of the CURRENT baseline provides additional benefit specifically for p_re recovery. This is Cp normalization COMBINED with asinh: predict `asinh(Cp * scale)` instead of `asinh(p * scale)`.

## Instructions

Change the pressure target normalization from raw pressure to Cp, combined with the existing asinh transform. The Cp denominator `q_inf = 0.5 * Umag²` varies per sample but is always positive and computable from existing input features.

### Implementation

1. **Read `prepare_multi.py` and `train.py`** to understand how pressure targets are currently normalized. Find where `asinh_pressure` and `asinh_scale` are applied.

2. **Modify the target normalization** for pressure channel only:
```python
# Current: target_p = asinh(p_raw * asinh_scale)
# New:     target_p = asinh(Cp * asinh_scale)
#          where Cp = p_raw / q_inf, q_inf = 0.5 * Umag^2

# q_inf computation (per sample):
umag = batch_global_features['Umag']  # or from input tensor
q_inf = 0.5 * umag ** 2  # [B] dynamic pressure
q_inf = q_inf.clamp(min=1e-6)  # avoid division by zero

# Normalize pressure target:
p_raw = target[:, :, pressure_channel]
Cp = p_raw / q_inf.unsqueeze(-1)  # [B, N] pressure coefficient
target_cp = torch.asinh(Cp * asinh_scale)

# Replace pressure target with Cp-based target
target[:, :, pressure_channel] = target_cp
```

3. **At inference/evaluation**: reverse the transformation to recover physical pressure for MAE computation:
```python
# Inverse: p_raw = sinh(model_output / asinh_scale) * q_inf
Cp_pred = torch.sinh(model_output_p) / asinh_scale  # recover Cp
p_pred = Cp_pred * q_inf  # recover raw pressure
```

4. **Important considerations:**
   - Umag must be available per-sample for both training and evaluation. Check how it's stored in the data pipeline.
   - Velocity targets (Ux, Uy) are NOT changed — only pressure.
   - The SRF head and residual prediction operate in the normalized space — they will now learn Cp residuals instead of p residuals.
   - At validation, the denormalized pressure MAE must be computed in raw pressure units for fair comparison with baseline.

5. **Add flag** `--cp_target_normalization`. When set, apply Cp normalization to pressure targets before asinh.

### Training command (seed 42)
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --seed 42 \
  --wandb_name "nezuko/cp-target-norm-s42" \
  --wandb_group "cp-target-normalization" \
  --cp_target_normalization \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

Seed 73: same with `--seed 73` and adjusted `--wandb_name`.

### What to report
1. 2-seed average for p_in, p_oodc, p_tan, p_re (in raw pressure MAE units, not Cp)
2. W&B run IDs
3. Whether p_re specifically improved (primary target of this experiment)
4. The range of q_inf values across the dataset (for understanding the normalization magnitude)

## Baseline (PR #2251, 2-seed avg)

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.891** | < 11.89 |
| **p_oodc** | **7.561** | < 7.56 |
| **p_tan** | **28.118** | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B baseline: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

**Question:** The current baseline code already applies Cp normalization through `_phys_norm()` (line 1213-1218).

The standard normalization path (active when neither `--adaptive_norm` nor `--raw_targets` is set, which is the baseline) does:
1. `y_phys = _phys_norm(y, Umag, q)` → computes Cp = p / q where q = 0.5 * Umag²
2. `y_phys[:, :, 2:3] = torch.asinh(y_phys[:, :, 2:3] * cfg.asinh_scale)` → asinh(Cp * 0.75)
3. Statistical normalization

So the model already predicts `asinh(Cp * 0.75)`, which is exactly what the PR instructions describe as the "new" target.

**Questions:**
1. Is the intent to change HOW Cp is computed? For example, using freestream velocity from input features (`flowState["Re"]`) instead of domain-mean velocity from `_umag_q()`?
2. Or should this be applied to the `adaptive_norm` path (which currently uses raw p instead of Cp)?
3. Or is there another interpretation I'm missing?

I want to make sure I implement the correct experiment before using GPU time. Please advise.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2289 because ADVISOR: Excellent catch — you're right, the baseline already computes Cp normalization through `_phys_norm()`. The experiment as designed is redundant with existing behavior.

This means the p_re regression (6.300→6.364) is NOT caused by missing Cp normalization — it's caused by something else, likely the cosine T_max=150 change tightening the learning rate schedule. The Cp normalization in `_phys_norm()` already handles the Re² pressure scaling.

Closing this PR. Assigning you a fresh experiment that targets p_re through a different mechanism: Re-stratified batch sampling..
